### PR TITLE
Controllers

### PR DESCRIPTION
**Pull request by brtkkndrl**
Implemented and updated controllers.
Changed FacadeBase. (Not updated in tests and facades yet, do it here or seperate commit, or wait for mergin of others?)
Changed DI Installer.

### CODE REVIEW

**Comment by KoubaFromPucks** [comment]
> Implemented and updated controllers. Changed FacadeBase. (Not updated in tests and facades yet, do it here or seperate commit, or wait for mergin of others?) Changed DI Installer.

In my opinion it is better to update it in next commit, but i would wait for @vecerap opinion.

File: src/Quizzer.API.App/Controllers/AnswerController.cs [hunk 1]
```diff
@@ -4,75 +4,81 @@
...
 
-        [HttpPut(Name = "UpdateAnswer")]
+        [HttpDelete("{id:guid}", Name = "[controller]/Delete")]
+        [ProducesResponseType(StatusCodes.Status404NotFound, Type = typeof(Guid))]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(Guid))]
+        public IActionResult Delete(Guid id) {
+            if (!_facade.Exists(id))
+                return NotFound(id);
+
+            _facade.Delete(id, DateTime.Now);
```
Comment (KoubaFromPucks) [inline src/Quizzer.API.App/Controllers/AnswerController.cs]: If the quiz has already started, an exception is thrown. So it means that API wil never notice a "requestor" that operation has failed.

File: src/Quizzer.API.App/Controllers/AnswerController.cs [hunk 2]
```diff
@@ -4,75 +4,81 @@
...
+            return Ok(id);
+        }
+
+        [HttpPut("{questionId:guid}", Name = "[controller]/Update")]
         [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(AnswerDetailModel))]
         [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(string))]
-        public IActionResult UpdateAnswer([FromBody] AnswerDetailModel answerModel, [FromBody] Guid questionId) {
+        public IActionResult Update([FromBody] AnswerDetailModel answerModel, Guid questionId) {
+            if (!_facade.Exists(questionId))
+                return BadRequest("Entity doesn't exist");
```
Comment (KoubaFromPucks) [inline src/Quizzer.API.App/Controllers/AnswerController.cs]: `_facade` is type of `AnswerFacade` so i guess you wanted to pass id of answer instead

File: src/Quizzer.API.App/Controllers/AnswerController.cs [hunk 3]
```diff
@@ -4,75 +4,81 @@
...
-
-        [HttpPost(Name = "CreateAnswer")]
+        [HttpPost("{questionId:guid}", Name = "[controller]/Create")]
         [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(Guid))]
         [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(string))]
-        public IActionResult CreateAnswer([FromBody] AnswerDetailModel answerModel, [FromBody] Guid questionId) {
-            Guid updatedId;
+        public IActionResult Create([FromBody] AnswerDetailModel answerModel, Guid questionId) {
+            if (_facade.Exists(questionId))
+                return BadRequest("Entity already exists");
```
Comment (KoubaFromPucks) [inline src/Quizzer.API.App/Controllers/AnswerController.cs]: Same as above - i guess you wanted to pass id of answer.

File: src/Quizzer.API.App/Controllers/QuizController.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        [HttpGet(Name = "[controller]/GetAll")]
+        public IEnumerable<QuizDetailModel> GetAll() {
+            return _facade.GetAll();
+        }
+
+        [HttpGet("{id:guid}", Name = "[controller]/GetById")]
+        [ProducesResponseType(StatusCodes.Status404NotFound, Type = typeof(Guid))]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(QuizDetailModel))]
+        public IActionResult GetById(Guid id) {
+            QuizDetailModel? user = _facade.GetDetailById(id);
```
Comment (KoubaFromPucks) [inline src/Quizzer.API.App/Controllers/QuizController.cs]: `user`?

File: src/Quizzer.API.App/Controllers/QuizController.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            if (user is null)
+                return NotFound(id);
+      
+            return Ok(user);
+        }
+
+        [HttpPut(Name = "[controller]/Update")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(QuizDetailModel))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(string))]
+        public IActionResult Update([FromBody] QuizDetailModel userModel) {
```
Comment (KoubaFromPucks) [inline src/Quizzer.API.App/Controllers/QuizController.cs]: `userModel`?

File: src/Quizzer.API.App/Controllers/QuizController.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+            } catch (Exception ex) {
+                return BadRequest(ex.Message);
+            }
+            return Ok(updatedId);
+        }
+
+        [HttpPost(Name = "[controller]/Create")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(Guid))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(string))]
+        public IActionResult Create([FromBody] QuizDetailModel userModel) {
```
Comment (KoubaFromPucks) [inline src/Quizzer.API.App/Controllers/QuizController.cs]: `userModel`?

File: src/Quizzer.API.App/Controllers/QuestionController.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        [HttpGet(Name = "[controller]/GetAll")]
+        public IEnumerable<QuestionDetailModel> GetAll() {
+            return _facade.GetAll();
+        }
+
+        [HttpGet("{id:guid}", Name = "[controller]/GetById")]
+        [ProducesResponseType(StatusCodes.Status404NotFound, Type = typeof(Guid))]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(QuestionDetailModel))]
+        public IActionResult GetById(Guid id) {
+            QuestionDetailModel? user = _facade.GetDetailById(id);
```
Comment (KoubaFromPucks) [inline src/Quizzer.API.App/Controllers/QuestionController.cs]: `user`?

File: src/Quizzer.API.App/Controllers/QuestionController.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            } catch (Exception ex) {
+                return BadRequest(ex.Message);
+            }
+            return Ok(updatedId);
+        }
+
+        [HttpPost("{quizId:guid}", Name = "[controller]/Create")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(Guid))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(string))]
+        public IActionResult Create([FromBody] QuestionDetailModel userModel, Guid quizId) {
```
Comment (KoubaFromPucks) [inline src/Quizzer.API.App/Controllers/QuestionController.cs]: `userModel`?

**Review by KoubaFromPucks — CHANGES_REQUESTED** [review]

**Comment by vecerap** [comment]
> > Implemented and updated controllers. Changed FacadeBase. (Not updated in tests and facades yet, do it here or seperate commit, or wait for mergin of others?) Changed DI Installer.
> 
> In my opinion it is better to update it in next commit, but i would wait for @vecerap opinion.

@BRTEK007 Which PRs would affect this work? I do not think any new will affect this, but I would separate it as well...

File: src/Quizzer.API.App/Controllers/UserController.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            if (!_facade.Exists(id))
+                return NotFound(id);
+
+            _facade.Delete(id);
+            return Ok(id);
+        }
+
+        [HttpGet("{name}", Name = "[controller]/GetByName")]
+        public IEnumerable<UserListModel> GetByName(string name) {
+            return _facade.GetByName(name, false);
```
Comment (vecerap) [inline src/Quizzer.API.App/Controllers/UserController.cs]: Why don't you have the option to be exact? you are hiding the functionality of the facade.

File: src/Quizzer.API.DAL/ServiceCollectionExtensions.cs
Comment (vecerap) [inline src/Quizzer.API.DAL/ServiceCollectionExtensions.cs:1]: Why did you remove this?

**Review by vecerap — COMMENTED** [review]
Which http request should we handle? The approach about the catching general exception is kind of vague. Now there is not much time to handle this properly, but I would consider making it more explicit later on...

File: src/Quizzer.API.App/Controllers/AnswerController.cs (hunk 1 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/AnswerController.cs]: updated

File: src/Quizzer.API.App/Controllers/AnswerController.cs (hunk 2 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/AnswerController.cs]: fixed

File: src/Quizzer.API.App/Controllers/AnswerController.cs (hunk 3 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/AnswerController.cs]: fixed

File: src/Quizzer.API.App/Controllers/QuizController.cs (hunk 4 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/QuizController.cs]: fixed

File: src/Quizzer.API.App/Controllers/QuizController.cs (hunk 5 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/QuizController.cs]: fixed

File: src/Quizzer.API.App/Controllers/QuizController.cs (hunk 6 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/QuizController.cs]: fixed

File: src/Quizzer.API.App/Controllers/QuestionController.cs (hunk 7 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/QuestionController.cs]: fixed

File: src/Quizzer.API.App/Controllers/QuestionController.cs (hunk 8 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/QuestionController.cs]: fixed

File: src/Quizzer.API.App/Controllers/UserController.cs (hunk 9 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/UserController.cs]: changed

File: src/Quizzer.API.DAL/ServiceCollectionExtensions.cs
Comment (brtkkndrl) [reply src/Quizzer.API.DAL/ServiceCollectionExtensions.cs:1]: I moved it to API.App since it was being used there. Now that I think shouldn't we move it to Quizzer.Extension? Or was there some reason to have it in multiple places?

**Comment by brtkkndrl** [comment]
> Which http request should we handle? The approach about the catching general exception is kind of vague. Now there is not much time to handle this properly, but I would consider making it more explicit later on...

If I understand this would require different types of exceptions being thrown in facade. Ideally there should also be a seperate naming table to strore all those error strings.

**Comment by brtkkndrl** [comment]
> > > Implemented and updated controllers. Changed FacadeBase. (Not updated in tests and facades yet, do it here or seperate commit, or wait for mergin of others?) Changed DI Installer.
> > 
> > 
> > In my opinion it is better to update it in next commit, but i would wait for @vecerap opinion.
> 
> @BRTEK007 Which PRs would affect this work? I do not think any new will affect this, but I would separate it as well...

Database fix, and facade tests, that have been merged already. I will split it facade changes to seperate commit now.

**Review by KoubaFromPucks — APPROVED** [review]

**Comment by vecerap** [comment]
> > Which http request should we handle? The approach about the catching general exception is kind of vague. Now there is not much time to handle this properly, but I would consider making it more explicit later on...
> 
> If I understand this would require different types of exceptions being thrown in facade. Ideally there should also be a seperate naming table to strore all those error strings.

Yeah the string table is kind of the extension of this project. What do you say about being more explicit when catching exceptions @KoubaFromPucks ?

**Review by vecerap — APPROVED** [review]
Squash commits

**Comment by vecerap** [comment]
@BRTEK007 Failing pipeline

**Comment by KoubaFromPucks** [comment]
> > > Which http request should we handle? The approach about the catching general exception is kind of vague. Now there is not much time to handle this properly, but I would consider making it more explicit later on...
> > 
> > 
> > If I understand this would require different types of exceptions being thrown in facade. Ideally there should also be a seperate naming table to strore all those error strings.
> 
> Yeah the string table is kind of the extension of this project. What do you say about being more explicit when catching exceptions @KoubaFromPucks ?

sound good

File: src/Quizzer.API.App/Controllers/QuestionController.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        [HttpGet("{id:guid}", Name = "[controller]/GetById")]
+        [ProducesResponseType(StatusCodes.Status404NotFound, Type = typeof(Guid))]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(QuestionDetailModel))]
+        public IActionResult GetById(Guid id) {
+            QuestionDetailModel? questionDetailModel = _facade.GetDetailById(id);
+
+            if (questionDetailModel is null)
+                return NotFound(id);
```
Comment (vecerap) [inline src/Quizzer.API.App/Controllers/QuestionController.cs]: do not use indentation

File: src/Quizzer.API.App/Controllers/QuestionController.cs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+            return Ok(questionDetailModel);
+        }
+
+        [HttpPut("{quizId:guid}", Name = "[controller]/Update")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(QuestionDetailModel))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(string))]
+        public IActionResult Update([FromBody] QuestionDetailModel questionDetailModel, Guid quizId) {
+            if (!_facade.Exists(questionDetailModel.Id))
+                return BadRequest("Entity doesn't exist");
```
Comment (vecerap) [inline src/Quizzer.API.App/Controllers/QuestionController.cs]: The same, either oneline or with braces

File: src/Quizzer.API.App/Controllers/QuestionController.cs [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+            }
+            return Ok(updatedId);
+        }
+
+        [HttpPost("{quizId:guid}", Name = "[controller]/Create")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(Guid))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(string))]
+        public IActionResult Create([FromBody] QuestionDetailModel questionDetailModel, Guid quizId) {
+            if (_facade.Exists(questionDetailModel.Id))
+                return BadRequest("Entity already exists");
```
Comment (vecerap) [inline src/Quizzer.API.App/Controllers/QuestionController.cs]: The same

File: src/Quizzer.API.App/Controllers/QuestionController.cs (hunk 10 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/QuestionController.cs]: fixed

File: src/Quizzer.API.App/Controllers/QuestionController.cs (hunk 11 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/QuestionController.cs]: fixed

File: src/Quizzer.API.App/Controllers/QuestionController.cs (hunk 12 above)
Comment (brtkkndrl) [reply src/Quizzer.API.App/Controllers/QuestionController.cs]: fixed
